champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/warlock/passives/Siphon.java

```diff
@@ -70,52 +98,151 @@ public double getEnergySiphoned(int level) {
         return baseEnergySiphoned + ((level - 1) * energySiphonedIncreasePerLevel);
     }
 
+    public double getSpeedDuration(int level) {
+        return speedDuration;
+    }
+
+    public double getElapsedTimeToProcAbility(int level) {
+        return elapsedTimeToProcAbility;
+    }
+
+    private long getRequiredSuccessfulUpdates(int level) {
+
+        // 1000ms in a second
+        double updatesPerSecond = ((double) 1000) / SIPHON_UPDATE_DELAY;
+
+        // Elapsed time is given in seconds
+        return Math.round( updatesPerSecond*getElapsedTimeToProcAbility(level) );
+    }
+
+    private double getRandomSiphonHealthGainChance(int level) {
+        return randomSiphonHealthGainChance;
+    }
+    private int getRandomSiphonHealthGainChanceAsPercentage(int level) {
+        return (int) Math.round(getRandomSiphonHealthGainChance(level)*100);
+    }
+
+    private double getHealthGainedOnRandomSiphon(int level) {
+        return healthGainedOnRandomSiphon;
+    }
+
+
     @Override
     public Role getClassType() {
         return Role.WARLOCK;
     }
 
-    @UpdateEvent(delay = 2000)
-    public void onUpdate() {
-        for (Player player : Bukkit.getOnlinePlayers()) {
+    @EventHandler
+    public void onEquip(SkillEquipEvent event) {
+        if (event.getSkill().equals(this)) {
+            active.add(event.getPlayer().getUniqueId());
+        }
+    }
+
+    @EventHandler
+    public void onDequip(SkillDequipEvent event) {
+        if (event.getSkill().equals(this)) {
+            active.remove(event.getPlayer().getUniqueId());
+        }
+    }
+
+    @UpdateEvent (delay = 500)
+    public void removeActives() {
+        playerUUIDToEnemyDataMap.keySet().forEach(playerUUID -> {
+            if (!active.contains(playerUUID)) playerUUIDToEnemyDataMap.remove(playerUUID);
+        });
+    }
+
+    @UpdateEvent(delay = SIPHON_UPDATE_DELAY)
+    public void monitorActives() {
+        active.forEach(playerUUID -> {
```

Comment:
I feel like you wouldn't a seperate active hashmap, just do it all within the playerUUIDToEnemyDataMap (maybe rename it to something less bad, like just siphonData or something

Reply:
Then iterate it with a standard iterator and remove entries if they dont have the skill etc

Reply:
I was able to condense the logic into a singular HashMap but do I need two UpdateEvents or is the way I'm doing things redundant?

Reply:
Just need one but use an iterator instead of forEach